framework/src/org/checkerframework/framework/flow/CFAbstractTransfer.java

```diff
@@ -721,12 +731,44 @@ protected void addInformationFromPreconditions(S info,
 
         S info = in.getRegularStore();
         V rhsValue = in.getValueOfSubNode(rhs);
+
+        Receiver expr = FlowExpressions.internalReprOf(analysis.getTypeFactory(),
+                n.getTarget());
+
+        if (inferSignatures && !expr.containsUnknown()
+                && expr instanceof FieldAccess) {
+            // Updates .jaif file
+            FieldAccessNode lhsNode = ((FieldAccessNode)lhs);
+            ClassSymbol clazzSymbol = SignatureInferenceScenes.getEnclosingClassSymbol(
+                    analysis.getContainingClass(n.getTree()), lhs,
+                    lhsNode.getReceiver());
+            SignatureInferenceScenes.updateFieldTypeInScene(
+                    lhsNode, rhs, clazzSymbol, analysis.getTypeFactory());
+        }
+
         processCommonAssignment(in, lhs, rhs, info, rhsValue);
 
         return new RegularTransferResult<>(finishValue(rhsValue, info), info);
     }
 
     @Override
+    public TransferResult<V, S> visitReturn(ReturnNode n, TransferInput<V, S> p) {
+        if (inferSignatures) {
+            // Updates the return type of the method on the respective .jaif file.
+            ClassTree classTree = analysis.getContainingClass(n.getTree());
+            if (classTree != null) {
```

Comment:
Under what circumstances could the class containing a return node be null?  Add a comment or remove the test.